3-Authorization-II/1-call-api/README.md

```diff
@@ -222,31 +234,140 @@ Were we successful in addressing your learning objective? Consider taking a mome
+app.use(cors());
+```
 
-### CORS Settings
+### Access token validation
 
-For the purpose of the sample, **cross-origin resource sharing** is enabled for **all** domains. This is insecure and not recommended. In production, you should modify this as to allow only the domains that you designate.
+On the web API side, [passport-azure-ad](https://github.com/AzureAD/passport-azure-ad) verifies the incoming access token's signature and validates it's payload against the `issuer` and `audience` claims (defined in `BearerStrategy` constructor) using the `passport.authenticate()` API. In the `BearerStrategy` callback, you can add further validation steps as shown below:
 
 ```javascript
-    app.use((req, res, next) => {
-        res.header("Access-Control-Allow-Origin", "*");
-        res.header("Access-Control-Allow-Headers", "Authorization, Origin, X-Requested-With, Content-Type, Accept");
-        next();
-    });
+const express = require('express');
+const passport = require('passport');
+const passportAzureAd = require('passport-azure-ad');
+
+const app = express();
+
+const bearerStrategy = new passportAzureAd.BearerStrategy({
+    identityMetadata: `https://${authConfig.metadata.authority}/${authConfig.credentials.tenantID}/${authConfig.metadata.version}/${authConfig.metadata.discovery}`,
+    issuer: `https://${authConfig.metadata.authority}/${authConfig.credentials.tenantID}/${authConfig.metadata.version}`,
+    clientID: authConfig.credentials.clientID,
+    audience: authConfig.credentials.clientID, // audience is this application
+    validateIssuer: authConfig.settings.validateIssuer,
+    passReqToCallback: authConfig.settings.passReqToCallback,
+    loggingLevel: authConfig.settings.loggingLevel
+}, (req, token, done) => {
+    /**
+     * Below you can do extended token validation and check for additional claims, such as:
+     * - check if the caller's tenant is in the allowed tenants list via the 'tid' claim (for multi-tenant applications)
+     * - check if the caller's account is homed or guest via the 'acct' optional claim
+     * - check if the caller belongs to right roles or groups via the 'roles' or 'groups' claim, respectively
+     *
+     * Bear in mind that you can do any of the above checks within the individual routes and/or controllers as well.
+     * For more information, visit: https://docs.microsoft.com/azure/active-directory/develop/access-tokens#validate-the-user-has-permission-to-access-this-data
+     */
+    if (requiredScopeOrAppPermission(token, [
+        ...authConfig.protectedRoutes.todolist.delegatedPermissions.read,
+        ...authConfig.protectedRoutes.todolist.delegatedPermissions.write,
+        ...authConfig.protectedRoutes.todolist.applicationPermissions.read,
+        ...authConfig.protectedRoutes.todolist.applicationPermissions.write,
```

Comment:
>...authConfig.protectedRoutes.todolist.applicationPermissions.write, [](http://example.com/codeflow?start=8&length=68)

add one example for "tid" validation as an example. It will not break the Single-tenant app and would allow for another example

Reply:
example added (to both README and code)